### PR TITLE
C22 - Phoenix - Tram Hoang

### PR DESCRIPTION
**Pull request by tramhn001**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool.isRequired,
+  onLikeToggle: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx]: The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`, as you did.

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,33 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeToggle }) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:5]: 👍 I like using destructured props to make it more visibly clear in the function definition itself what props we're expecting to receive. We do need to remember that these are all passed in as a single object (the one we usually call `props`) and it can cause problems if we forget to include the destructuring syntax (it's easy to forget and list the props as multiple separate params) I really prefer the glanceability.

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,33 @@
...
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={`chat-entry ${liked ? 'liked' : ''} ${sender === 'Vladimir' ? 'local' : 'remote'}`}>
+      <h2 className="entry-name">{sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:17]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: dist/assets/index-CkmkyOUd.css
Comment (anselrognlie) [inline dist/assets/index-CkmkyOUd.css:1]: It should be possible to add `dist` to the `.gitignore` so the `dist` folder doesn't get checked in to your code branch. Build products generally shouldn't get checked into the source repo.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.string.isRequired,
+      liked: PropTypes.bool.isRequired
+    })
+  ).isRequired,
+  onToggleLike: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import ChatEntry from './ChatEntry.jsx';
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+
+const ChatLog = ({ entries, onToggleLike }) => {
+	return (
+		<div className="chat-log">
+			{entries.map((entry) => (
+				<ChatEntry
+          key={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:10]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. Generally we would also pass the `id` into our component even though it might feel redundant to pass both, but one is for our own logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import ChatEntry from './ChatEntry.jsx';
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+
+const ChatLog = ({ entries, onToggleLike }) => {
+	return (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nit: This `return` and its expression seems very deeply indented. Maintain a consistent indentation, and avoid unindenting a portion of the expression (the `ChatEntry` props).

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+		<div className="chat-log">
+			{entries.map((entry) => (
+				<ChatEntry
+          key={entry.id}
+          sender={entry.sender}
+          body={entry.body}
+          timeStamp={entry.timeStamp}
+          liked={entry.liked}
+          onLikeToggle={() => onToggleLike(entry.id)}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx]: Prefer to pass in the `id` on its own and just pass through the `onToggleLike` function received through the props.
```jsx
          onLikeToggle={onToggleLike}
```
Try to settle on a style for naming events. We see two approaches here. `onLikeToggle` is of the form target-action, while `onToggleLike` is structured as action-target.

File: src/components/ChatLog.jsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import ChatEntry from './ChatEntry.jsx';
+import PropTypes from 'prop-types';
+import './ChatLog.css';
+
+const ChatLog = ({ entries, onToggleLike }) => {
+	return (
+		<div className="chat-log">
+			{entries.map((entry) => (
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:8]: Nice use of `map` to convert from the message data into `ChatEntry` components. We could perform this mapping storing the result into a variable we use in the JSX result, we could make a helper function that we call as part of the `return`, or we can perform the map expression as part of the `return` JSX as you did here. I often like having relatively simple maps directly in the JSX since it helps me see the overall structure of the component, though it can make debugging a little more tricky. But any of those approaches will work fine.

File: package.json [hunk 9]
```diff
@@ -2,8 +2,11 @@
   "name": "react-chatlog",
   "private": true,
   "version": "0.0.0",
+  "homepage": "https://tramhn001.github.io/react-chatlog/",
```
Comment (anselrognlie) [inline package.json:5]: Since you set the `base` in `vite.config.js` it shouldn't be necessary to set `homepage` here. I'm not sure what effect if any this has for a vite build.

File: package.json [hunk 10]
```diff
@@ -2,8 +2,11 @@
...
   "type": "module",
   "scripts": {
+    "predeploy": "npm run build",
+    "deploy": "gh-pages -d build",
```
Comment (anselrognlie) [inline package.json:9]: This can be set to `gh-pages -d dist` to look in the output directory vite is using. This would prevent needing to tweak the `outDir` in `vite.config.js`.

If possible, I tend to leave the build tool (vite in this case) closer to its own default behaviors and then tweak the downstream tools to match rather than making vite conform to the `gh-pages` assumptions.

File: vite.config.js [hunk 11]
```diff
@@ -4,6 +4,10 @@ import react from '@vitejs/plugin-react'
 // https://vitejs.dev/config/
 export default defineConfig({
   plugins: [react()],
+  build: {
+    outDir: 'build', // Change 'dist' to 'build' for compatibility with gh-pages
+  },
```
Comment (anselrognlie) [inline vite.config.js:9]: This change is unnecessary if we tell `gh-pages` to look at the `dist` directory.

File: src/App.jsx [hunk 12]
```diff
@@ -1,14 +1,34 @@
 import './App.css';
+import messages from './data/messages.json';
+import ChatLog from './components/ChatLog';
+import { useState } from 'react';
 
 const App = () => {
+  const [chatMessages, setChatMessages] = useState(messages);
+
+  const toggleLike = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:9]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 13]
```diff
@@ -1,14 +1,34 @@
...
+    const updatedMessages = chatMessages.map((message) =>
```
Comment (anselrognlie) [inline src/App.jsx]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 14]
```diff
@@ -1,14 +1,34 @@
...
+      message.id === id ? { ...message, liked: !message.liked } : message
```
Comment (anselrognlie) [inline src/App.jsx]: Opinions can definintely vary here, but for me, this is starting down the road of doing too much for a ternary. I like my ternaries to be very glanceable, but here, with the new object syntax, it gets a little hard to read. It also trivializes the fact that yes we are building a whole new object here and applying the like logic. As a result, I would tend to stick with an `if`/`else`.

File: src/App.jsx [hunk 15]
```diff
@@ -1,14 +1,34 @@
...
+    );
+    setChatMessages(updatedMessages);
```
Comment (anselrognlie) [inline src/App.jsx:18]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setChatMessages(chatMessages => chatMessages.map((message) =>
      message.id === id ? { ...message, liked: !message.liked } : message
    ));
```

File: src/App.jsx [hunk 16]
```diff
@@ -1,14 +1,34 @@
...
+  };
+
+  const totalLikes = chatMessages.filter((message) => message.liked).length;
```
Comment (anselrognlie) [inline src/App.jsx:21]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

Great idea to filter down the data to the liked messages and use the list length as the count. This is a really understandable alternative to the more canonical `reduce` approach. Compared to `reduce`, it does make a list potentially as long as the message list, but since we know a similar copying gets done anyway during a render (building the `ChatEntry` components) on the whole it's not any more costly.

File: src/components/ChatEntry.jsx [hunk 17]
```diff
@@ -1,20 +1,33 @@
...
+        </p>
+        <button 
+          className="like" 
+          onClick={() => onLikeToggle(id)}>
+          {liked ? '❤️' : '🤍'}
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:22]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 18]
```diff
@@ -1,20 +1,33 @@
...
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{body}</p>
+        <p className="entry-time">
+          <TimeStamp time={timeStamp} />
+        </p>
+        <button 
+          className="like" 
+          onClick={() => onLikeToggle(id)}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:21]: 👀 From this code, it looks like you were planning to have `onLikeToggle` be afunction that accepts an id as a parameter, and that the `id` of this message would be passed. However, because `ChatLog` did not pass down the  message id, the `id` always ends up being `undefined`. `ChatLog` also creates a per-`ChatEntry` callback function wrapping the like toggle function passed from `App`, and that's what's actually being called here. That function doesn't take any params, so the (`undefined`) `id` being passed here is currently being ignored.

File: src/components/ChatLog.jsx (hunk 5 above)
Comment (anselrognlie) [reply src/components/ChatLog.jsx:10]: 👀 Note that your `ChatEntry` prop types marked the `id` as a required prop, and further, the way that the like toggle function is used within `ChatEntry` implies that you really were intending to pass the `id` in. It looks like you created a work around for this by creating a custom like toggle function per instance that captures the `id`, but conceptually this moves some of the responsibility for knowing how to wire up logic internal to `ChatEntry` here in `ChatLog`. Rather than building a new function here, it would be more typical to pass the `id` all the way into `ChatEntry`.

File: src/components/ChatEntry.jsx [hunk 19]
```diff
@@ -1,20 +1,33 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
 
-const ChatEntry = () => {
+const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeToggle }) => {
   return (
-    <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+    <div className={`chat-entry ${liked ? 'liked' : ''} ${sender === 'Vladimir' ? 'local' : 'remote'}`}>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:12]: Nice logic to decide whether to treat a message as local or remote. How could we generalize this so that it didn't need to look explicitly for Vladimir? This project only passes in a single conversation, but ideally, our components should work in other situations.

In the general case, does the `ChatEntry` itself have enough information as it is to "know" which messages are local and which are remote?

File: src/App.jsx [hunk 20]
```diff
@@ -1,14 +1,34 @@
...
+
+  const sender1 = messages[0].sender // First sender
+  const sender2 = messages[1].sender // Second sender
```
Comment (anselrognlie) [inline src/App.jsx]: Nice way to read the participant names from the messages. Notice that this assumes there are only two participants in this conversation, and that they are found in the first two messages. What would happen for other conversation situations?
- no one has sent a message yet
- only one participant has sent a message
- a single participant sends multiple messages before getting a response
- there are more than two participants in this conversation
Some of these cases might not really be workable given the limited data we're working with, but it's worth thinking about what this could look like for a more complete application.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments as there are a few things I think it's worth taking a closer look at. Let me know if you have any questions.

File: src/components/ChatEntry.jsx (hunk 1 above)
Comment (tramhn001) [reply src/components/ChatEntry.jsx:30]: Can I fix it like this:
ChatEntry.propTypes = {
  id: PropTypes.number.isRequired,
  sender: PropTypes.string.isRequired,
  body: PropTypes.string.isRequired,
  timeStamp: PropTypes.string.isRequired,
  liked: PropTypes.bool.isRequired,
  onLikeToggle: PropTypes.func,
};

ChatEntry.defaultProps = {
  onLikeToggle: () => {}, // Default to an empty function
}

File: dist/assets/index-CkmkyOUd.css
Comment (anselrognlie) [reply dist/assets/index-CkmkyOUd.css:1]: We'd want the _entire_ `dist` folder excluded, not just this single file. Note that it's also a bit cumbersome to clean the files out of the repository. Before ignoring `dist`, you'd want to delete the `dist` folder, add, commit, and push that change. _Then_ you would add `dist` to the `.gitignore`. If a file is listed in `.gitignore`, git won't watch for changes on it any more, so it needs to be removed from the `.gitignore` (add, commit, push) before making any changes.

File: src/App.jsx [hunk 21]
```diff
@@ -7,21 +7,26 @@ const App = () => {
...
+
     setChatMessages(updatedMessages);
   };
 
   const totalLikes = chatMessages.filter((message) => message.liked).length;
 
-  const sender1 = messages[0].sender // First sender
-  const sender2 = messages[1].sender // Second sender
+  // Extract unique participants
+  const participants = [...new Set(chatMessages.map((message) => message.sender))];
```
Comment (anselrognlie) [inline src/App.jsx:24]: Nice use of `map` to extract the sender information, then uniquing that with a `Set` to get the participants in the conversation. We can confirm with the docs that iterating over the `Set` contents will occur in insertion order, meaning the resulting list of senders, however many there are, will be in the order they participated in the conversation.

File: src/App.jsx [hunk 22]
```diff
@@ -7,21 +7,26 @@ const App = () => {
...
 
   return (
     <div id="App">
       <header>
-        <h1>{`Chat between ${sender1} and ${sender2}`}</h1>
+        <h1>{`Chat between ${participants.join(' and ')}`}</h1>
```
Comment (anselrognlie) [inline src/App.jsx:29]: If there were more than 2 participants we'd probably want some additional logic, but this works great for a typical 2-person conversation.

File: src/components/ChatEntry.jsx [hunk 23]
```diff
@@ -27,7 +32,7 @@ ChatEntry.propTypes = {
   body: PropTypes.string.isRequired,
   timeStamp: PropTypes.string.isRequired,
   liked: PropTypes.bool.isRequired,
-  onLikeToggle: PropTypes.func.isRequired,
+  onLikeToggle: PropTypes.func,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:35]: 👍 Removing the `isRequired` marker avoids the prop type warning.

File: src/components/ChatEntry.jsx [hunk 24]
```diff
@@ -3,6 +3,11 @@ import TimeStamp from './TimeStamp';
 import PropTypes from 'prop-types';
 
 const ChatEntry = ({ id, sender, body, timeStamp, liked, onLikeToggle }) => {
+  const handleLikeClick = () => {
+    if (onLikeToggle) {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 Only calling the function if it's supplied ensures that our code doesn't try to use a value that isn't supplied, which is possible since we no longer mark it required.

File: src/components/ChatLog.jsx [hunk 25]
```diff
@@ -2,24 +2,25 @@ import ChatEntry from './ChatEntry.jsx';
 import PropTypes from 'prop-types';
 import './ChatLog.css';
 
-const ChatLog = ({ entries, onToggleLike }) => {
+const ChatLog = ({ entries, onToggleLike = () => {} }) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:5]: 👍 Providing a default value is another useful strategy for dealing non-required props. Though in this case, consider hvaing the function still recieve the same expected params as a "real" function to help a reader see what data the function will have to work with (the id).

File: src/components/ChatLog.jsx [hunk 26]
```diff
@@ -2,24 +2,25 @@ import ChatEntry from './ChatEntry.jsx';
...
 	return (
 		<div className="chat-log">
 			{entries.map((entry) => (
 				<ChatEntry
           key={entry.id}
+          id={entry.id} // Pass the id directly
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:11]: 👍 Passing the id all the way into the `ChatEntry` component lets it know which message it's working with.

File: src/components/ChatLog.jsx [hunk 27]
```diff
@@ -2,24 +2,25 @@ import ChatEntry from './ChatEntry.jsx';
...
           sender={entry.sender}
           body={entry.body}
           timeStamp={entry.timeStamp}
           liked={entry.liked}
-          onLikeToggle={() => onToggleLike(entry.id)}
+          onLikeToggle={onToggleLike} // Pass the function directly
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:16]: 👍 We can drill the props down to the individual `ChatEntry` components rather than building a new function here. `ChatEntry` still makes its own per-component function `handleLikeClick`, so we're not doing this out of concern for the resources being consumed by the function, but rather fomr the perspective of which component should know how to wire up the function. In this case, `ChatLog`'s responsibility is passing data through to the individual entries, and the entries have the responsibility of knowing how to use the values.

File: src/components/ChatLog.jsx [hunk 28]
```diff
@@ -2,24 +2,25 @@ import ChatEntry from './ChatEntry.jsx';
...
 				/>
 			))}
 		</div>
 	);
 };
 
-ChatLog.PropTypes = {
+ChatLog.propTypes = {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:23]: 👍 Fixing the typo allows these prop types to be checked.

**Review by anselrognlie — COMMENTED** [review]
Thanks for the updates. This all looks good!
